bookkeeper-server/src/main/java/org/apache/bookkeeper/http/RecoveryBookieService.java

```diff
@@ -0,0 +1,133 @@
+
+    protected ServerConfiguration conf;
+
+    public RecoveryBookieService(ServerConfiguration conf) {
+        Preconditions.checkNotNull(conf);
+        this.conf = conf;
+    }
+
+    /*
+     * Example body as this:
+     * {
+     *   "bookie_src": [ "bookie_src1", "bookie_src2"... ],
+     *   "bookie_dest": [ "bookie_dest1", "bookie_dest2"... ],
+     *   "delete_cookie": <bool_value>
+     * }
+     */
+    static class RecoveryRequestJsonBody {
+        public List<String> bookie_src;
+        public List<String> bookie_dest;
+        public boolean delete_cookie;
+
+        /*public RecoveryRequestJsonBody () {
+            bookie_src = null;
+            bookie_dest = null;
+            delete_cookie = false;
+        }*/
+    }
+
+    @Override
+    public HttpServiceResponse handle(HttpServiceRequest request) throws Exception {
+        HttpServiceResponse response = new HttpServiceResponse();
+        String requestBody = request.getBody();
+        RecoveryRequestJsonBody requestJsonBody;
+
+        if (requestBody == null) {
+            response.setCode(HttpServer.StatusCode.NOT_FOUND);
+            response.setBody("No request body provide.");
+            return response;
+        }
+
+        try {
+            requestJsonBody = JsonUtil.fromJson(requestBody, RecoveryRequestJsonBody.class);
+            LOG.debug("bookie_src: [" + requestJsonBody.bookie_src.get(0)
+                + "],  bookie_dest: ["
+                + ((requestJsonBody.bookie_dest == null) ? "null" : requestJsonBody.bookie_dest.get(0))
+                + "],  delete_cookie: [" + requestJsonBody.delete_cookie + "]");
+        } catch (JsonUtil.ParseJsonException e) {
+            LOG.error("Meet Exception: ", e);
+            response.setCode(HttpServer.StatusCode.NOT_FOUND);
+            response.setBody("ERROR parameters: " + e.getMessage());
+            return response;
+        }
+
+        if (HttpServer.Method.PUT == request.getMethod() &&
+            !requestJsonBody.bookie_src.isEmpty()) {
+            ClientConfiguration adminConf = new ClientConfiguration(conf);
+            BookKeeperAdmin admin = new BookKeeperAdmin(adminConf);
+
+            String bookieSrcString[] = requestJsonBody.bookie_src.get(0).split(":");
+            BookieSocketAddress bookieSrc = new BookieSocketAddress(
+              bookieSrcString[0], Integer.parseInt(bookieSrcString[1]));
+            BookieSocketAddress bookieDest = null;
+            if ((requestJsonBody.bookie_dest != null) && !requestJsonBody.bookie_dest.isEmpty()) {
+                String bookieDestString[] = requestJsonBody.bookie_dest.get(0).split(":");
+                bookieDest = new BookieSocketAddress(bookieDestString[0],
+                  Integer.parseInt(bookieDestString[1]));
+            }
+            boolean deleteCookie = requestJsonBody.delete_cookie;
+            try {
+                admin.recoverBookieData(bookieSrc, bookieDest);
```

Comment:
Looks to me another long running process, same issue as mentioned before

Reply:
Got it.